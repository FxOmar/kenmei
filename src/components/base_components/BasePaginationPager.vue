<template lang="pug">
  nav.relative.z-0.inline-flex.shadow-sm.rounded-l-md
    a.px-2.py-1.rounded-l-md.text-gray-500.hover_text-gray-400.border-l(
      href='javascript:void(0)'
      aria-label='Previous'
      :class="{ 'cursor-not-allowed': !pagy.prev }"
      @click.prevent="changePage(pagy.prev)"
    )
      icon-chevron-left.h-5.w-5(:class="{ 'opacity-50': !pagy.prev }")
    template(v-for="page in pagy.series")
      a.-ml-px.px-4.py-1.text-gray-700(
        href='javascript:void(0)'
        v-if="page !== 'gap'"
        v-text="page"
        @click.prevent="changePage(page)"
        :class="paginationClasses(page)"
      )
      span.gap(v-else)
        | â€¦
    a.-ml-px.px-2.py-1.rounded-r-md.text-gray-500.hover_text-gray-400.border-r(
      href='javascript:void(0)'
      aria-label='Next'
      :class="{ 'cursor-not-allowed': !pagy.next }"
      @click.prevent="changePage(pagy.next)"
    )
      icon-chevron-right.h-5.w-5(:class="{ 'opacity-50': !pagy.next }")
</template>

<script>
  export default {
    props: {
      pagy: {
        type: Object,
        required: true,
      },
      loading: {
        type: Boolean,
        default: false,
      },
    },
    data() {
      return {
        currentPage: null,
      };
    },
    computed: {
      currentPagyPage() {
        const page = this.pagy.series.find(
          (page) => !Number.isInteger(page) && page !== 'gap',
        );

        return Number.parseInt(page, 10);
      },
    },
    mounted() {
      this.currentPage = this.currentPagyPage;
    },
    updated() {
      if (this.loading || this.currentPagyPage === this.currentPage) return;

      this.currentPage = this.currentPagyPage;
    },
    methods: {
      changePage(page) {
        if (!page) return;

        this.$emit('pageChanged', page);
        this.currentPage = page;
      },
      isCurrentPage(page) {
        return Number.parseInt(page, 10) === this.currentPage;
      },
      paginationClasses(page) {
        return {
          'text-blue-500 pointer-events-none': this.isCurrentPage(page),
          'text-gray-500 hover_text-gray-400': !this.isCurrentPage(page),
        };
      },
    },
  };
</script>

<style lang="scss" scoped>
  a {
    @apply relative inline-flex items-center border-t border-b;
    @apply border-gray-300 bg-white text-sm leading-5 font-medium;
    @apply transition ease-in-out duration-150;
  }

  .gap {
    @apply -ml-px relative inline-flex items-center px-4 py-1 border-t border-b;
    @apply border-gray-300 bg-white text-sm leading-5 font-medium text-gray-700;
  }
</style>
