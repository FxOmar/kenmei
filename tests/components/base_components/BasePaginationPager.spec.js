import BasePaginationPager from '@/components/base_components/BasePaginationPager.vue';

describe('BasePaginationPager.vue', () => {
  let paginationPager;

  beforeEach(() => {
    paginationPager = shallowMount(BasePaginationPager, {
      propsData: {
        pagy: {
          from: 1, to: 2, count: 2, series: ['1', 2], prev: null, next: 2,
        },
      },
    });
  });

  it('shows gap page', () => {
    expect(paginationPager.html()).toMatchSnapshot();
  });

  describe('when gap is present', () => {
    beforeEach(() => {
      paginationPager.setProps({
        pagy: {
          from: 1, to: 2, count: 2, series: ['1', 'gap', 2], prev: null, next: 2,
        },
      });
    });

    it('shows gap page', () => {
      expect(paginationPager.text()).toContain('â€¦');
    });
  });

  describe('when previous button is not in pagy', () => {
    it('disables it', () => {
      expect(paginationPager.findAll('a').at(0).element)
        .toHaveClass('cursor-not-allowed');
    });
  });

  describe('when next button is not in pagy', () => {
    beforeEach(() => {
      paginationPager.setProps({
        pagy: {
          from: 1, to: 2, count: 2, series: [1, '2'], prev: 1, next: null,
        },
      });
    });

    it('disables it', () => {
      expect(paginationPager.findAll('a').at(3).element)
        .toHaveClass('cursor-not-allowed');
    });
  });

  describe('when component is mounted', () => {
    it('sets currentPage to currentPagyPage', () => {
      expect(paginationPager.vm.$data.currentPage).toEqual(1);
    });
  });

  describe('when component is updated', () => {
    describe('and pagy has a new current page', () => {
      beforeEach(() => {
        paginationPager.setProps({
          pagy: {
            from: 1, to: 2, count: 2, series: [1, '2'], prev: 1, next: null,
          },
          loading: false,
        });
      });

      it('sets currentPage to currentPagyPage', () => {
        expect(paginationPager.vm.$data.currentPage).toEqual(2);
      });
    });

    describe('and loading is true', () => {
      beforeEach(() => {
        paginationPager.setProps({
          pagy: {
            from: 1, to: 2, count: 2, series: [1, '2'], prev: 1, next: null,
          },
          loading: true,
        });
      });

      it("doesn't update currentPage", () => {
        expect(paginationPager.vm.$data.currentPage).toEqual(1);
      });
    });

    describe('and currentPage equals currentPagyPage', () => {
      beforeEach(() => {
        paginationPager.setData({ currentPage: 2 });
        paginationPager.setProps({
          pagy: {
            from: 1, to: 2, count: 2, series: [1, '2'], prev: 1, next: null,
          },
          loading: false,
        });
      });

      it("doesn't update currentPage", () => {
        expect(paginationPager.vm.$data.currentPage).toEqual(2);
      });
    });
  });

  describe('when changing pages', () => {
    it('emits pageChanged and sets currentPage', async () => {
      await paginationPager.findAll('a').at(3).trigger('click');

      expect(paginationPager.emitted('pageChanged')[0]).toEqual([2]);
    });

    describe('and no page provided', () => {
      it('does nothing', () => {
        paginationPager.vm.changePage(null);

        expect(paginationPager.emitted('pageChanged')).toBeFalsy();
      });
    });
  });
});
