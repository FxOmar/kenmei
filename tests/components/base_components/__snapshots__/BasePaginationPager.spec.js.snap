// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BasePaginationPager.vue shows gap page 1`] = `
<nav class="relative z-0 inline-flex shadow-sm rounded-l-md"><a href="javascript:void(0)" aria-label="Previous" class="px-2 py-1 rounded-l-md text-gray-500 hover_text-gray-400 border-l cursor-not-allowed">
    <icon-chevron-left-stub class="h-5 w-5 opacity-50"></icon-chevron-left-stub>
  </a><a href="javascript:void(0)" class="-ml-px px-4 py-1 text-gray-700 text-blue-500 pointer-events-none">1</a><a href="javascript:void(0)" class="-ml-px px-4 py-1 text-gray-700 text-gray-500 hover_text-gray-400">2</a><a href="javascript:void(0)" aria-label="Next" class="-ml-px px-2 py-1 rounded-r-md text-gray-500 hover_text-gray-400 border-r">
    <icon-chevron-right-stub class="h-5 w-5"></icon-chevron-right-stub>
  </a></nav>
`;
